Drop IsDeprecatedWeakRefSmartPointerException from SQLiteStorageArea & StorageAreaBase

cdumez opened the pull request:
#### f1f97ea13f5c953bd909ce22ae07a87bf331da1e
<pre>
Drop IsDeprecatedWeakRefSmartPointerException from SQLiteStorageArea &amp; StorageAreaBase
<a href="https://bugs.webkit.org/show_bug.cgi?id=281393">https://bugs.webkit.org/show_bug.cgi?id=281393</a>

Reviewed by Darin Adler.

* Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
(WebKit::LocalStorageManager::hasDataInMemory const):
(WebKit::LocalStorageManager::clearDataInMemory):
(WebKit::LocalStorageManager::clearDataOnDisk):
(WebKit::LocalStorageManager::close):
(WebKit::LocalStorageManager::handleLowMemoryWarning):
(WebKit::LocalStorageManager::syncLocalStorage):
(WebKit::LocalStorageManager::connectionClosedForLocalStorageArea):
(WebKit::LocalStorageManager::connectionClosedForTransientStorageArea):
(WebKit::LocalStorageManager::connectToLocalStorageArea):
(WebKit::LocalStorageManager::protectedLocalStorageArea const):
(WebKit::LocalStorageManager::connectToTransientLocalStorageArea):
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.h:
* Source/WebKit/NetworkProcess/storage/MemoryStorageArea.cpp:
(WebKit::MemoryStorageArea::create):
(WebKit::MemoryStorageArea::clone const):
* Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h:
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
(WebKit::SQLiteStorageArea::create):
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.h:
* Source/WebKit/NetworkProcess/storage/SessionStorageManager.cpp:
(WebKit::SessionStorageManager::clearData):
(WebKit::SessionStorageManager::connectionClosed):
(WebKit::SessionStorageManager::addStorageArea):
(WebKit::SessionStorageManager::connectToSessionStorageArea):
(WebKit::SessionStorageManager::cancelConnectToSessionStorageArea):
(WebKit::SessionStorageManager::disconnectFromStorageArea):
(WebKit::SessionStorageManager::cloneStorageArea):
* Source/WebKit/NetworkProcess/storage/SessionStorageManager.h:
* Source/WebKit/NetworkProcess/storage/StorageAreaBase.cpp:
(WebKit::StorageAreaBase::hasListeners const): Deleted.
* Source/WebKit/NetworkProcess/storage/StorageAreaBase.h:
(WebKit::StorageAreaBase::hasListeners const):

Canonical link: <a href="https://commits.webkit.org/285106@main">https://commits.webkit.org/285106@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9845132c41533e65095ab7fcfb9879ffc6dac35a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71573 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50986 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24347 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75685 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22778 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58787 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22598 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56528 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15001 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74639 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46257 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61649 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36977 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42920 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21119 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64823 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19479 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77403 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15807 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18656 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64244 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15850 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/61688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64239 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12384 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6023 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10968 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46786 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/1565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47857 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49141 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47599 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
```diff
@@ -107,40 +107,44 @@ bool LocalStorageManager::isActive() const
+    if (memoryLocalStorage && !memoryLocalStorage->isEmpty())
+        return true;
+
+    RefPtr transientStorageArea = m_transientStorageArea;
```
To match the other local variable names, this would just be `transientStorage`.

darinadler commented on Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
```diff
@@ -107,40 +107,44 @@ bool LocalStorageManager::isActive() const
 
-    if (m_transientStorageArea)
-        m_transientStorageArea->clear();
+    if (RefPtr transientStorageArea = m_transientStorageArea)
```
To match the other local variable names, this would just be `transientStorage`. Shorter is likely better, but also more consistent would be nicer. Not really important either way. If I was writhing the code, in a class named "storage manager" I’d use the word "area" rather than "storage", since "storage" is implied by the class.

darinadler commented on Source/WebKit/NetworkProcess/storage/StorageAreaBase.cpp:
```diff
@@ -59,10 +59,7 @@ void StorageAreaBase::removeListener(IPC::Connection::UniqueID connection)
-{
-    return !m_listeners.isEmpty();
-}
+
```
Shouldn't leave an extra blank line.

darinadler commented on Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h:
```diff
@@ -35,18 +35,22 @@ class StorageMap;
-    std::unique_ptr<MemoryStorageArea> clone() const;
+    Ref<MemoryStorageArea> clone() const;
     
+    void ref() const final { RefCounted::ref(); }
+    void deref() const final { RefCounted::deref(); }
```
I believe this can be done with `using`, and doing so might even have a tiny efficiency effect.
```
using RefCounted::ref;
using RefCounted::deref;
```
I found that [cppreference.com](https://en.cppreference.com/w/cpp/language/using_declaration) says, "If the derived class already has a member with the same name, parameter list, and qualifications, the derived class member hides or overrides (doesn't conflict with) the member that is introduced from the base class."

I also noticed that we have macros named `DECLARE_VIRTUAL_REFCOUNTED` and `DEFINE_VIRTUAL_REFCOUNTED` that are apparently for this idiom. Maybe we should be using those in this pull request?

If so, then maybe later we can change them to the `using` technique.

cdumez replied:
> I also noticed that we have macros named DECLARE_VIRTUAL_REFCOUNTED and DEFINE_VIRTUAL_REFCOUNTED that are apparently for this idiom. Maybe we should be using those in this pull request?

I am discussing those macros with Geoff (who introduced them) because I am not in favor of using them. I'd rather not hide ref() / deref() functions behind an obscure macro. This is also only 2 trivial lines so it doesn't feel worth the decreased readability / understandability from using the macro.

I'll look into `using` statements for this, I didn't realize this would work.

cdumez replied:
Your proposal didn't build. I made this change:
```
diff --git a/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h b/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h
index 098a2d8d64b9..a2ecba929a65 100644
--- a/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h
+++ b/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h
@@ -45,8 +45,8 @@ public:
     void clear() final;
     Ref<MemoryStorageArea> clone() const;
     
-    void ref() const final { RefCounted::ref(); }
-    void deref() const final { RefCounted::deref(); }
+    using RefCounted::ref;
+    using RefCounted::deref;
 
 private:
     explicit MemoryStorageArea(const WebCore::ClientOrigin&, StorageAreaBase::StorageType);
```

and got:
```
/Volumes/Work/WebKit/OpenSource/Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h:38:7: error: abstract class is marked 'final' [-Werror,-Wabstract-final-class]
   38 | class MemoryStorageArea final : public StorageAreaBase, public RefCounted<MemoryStorageArea> {
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/unified-sources/UnifiedSource36.cpp:1:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:27:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.h:28:
/Volumes/Work/WebKit/OpenSource/Source/WebKit/NetworkProcess/storage/StorageAreaBase.h:76:18: note: unimplemented pure virtual method 'ref' in 'MemoryStorageArea'
   76 |     virtual void ref() const = 0;
/Volumes/Work/WebKit/OpenSource/Source/WebKit/NetworkProcess/storage/StorageAreaBase.h:77:18: note: unimplemented pure virtual method 'deref' in 'MemoryStorageArea'
   77 |     virtual void deref() const = 0;
```

darinadler replied:
Darn, I was wrong. Thanks for trying it.

cdumez committed:
Drop IsDeprecatedWeakRefSmartPointerException from SQLiteStorageArea & StorageAreaBase
https://bugs.webkit.org/show_bug.cgi?id=281393

Reviewed by Darin Adler.

* Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
(WebKit::LocalStorageManager::hasDataInMemory const):
(WebKit::LocalStorageManager::clearDataInMemory):
(WebKit::LocalStorageManager::clearDataOnDisk):
(WebKit::LocalStorageManager::close):
(WebKit::LocalStorageManager::handleLowMemoryWarning):
(WebKit::LocalStorageManager::syncLocalStorage):
(WebKit::LocalStorageManager::connectionClosedForLocalStorageArea):
(WebKit::LocalStorageManager::connectionClosedForTransientStorageArea):
(WebKit::LocalStorageManager::connectToLocalStorageArea):
(WebKit::LocalStorageManager::protectedLocalStorageArea const):
(WebKit::LocalStorageManager::connectToTransientLocalStorageArea):
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.h:
* Source/WebKit/NetworkProcess/storage/MemoryStorageArea.cpp:
(WebKit::MemoryStorageArea::create):
(WebKit::MemoryStorageArea::clone const):
* Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h:
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
(WebKit::SQLiteStorageArea::create):
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.h:
* Source/WebKit/NetworkProcess/storage/SessionStorageManager.cpp:
(WebKit::SessionStorageManager::clearData):
(WebKit::SessionStorageManager::connectionClosed):
(WebKit::SessionStorageManager::addStorageArea):
(WebKit::SessionStorageManager::connectToSessionStorageArea):
(WebKit::SessionStorageManager::cancelConnectToSessionStorageArea):
(WebKit::SessionStorageManager::disconnectFromStorageArea):
(WebKit::SessionStorageManager::cloneStorageArea):
* Source/WebKit/NetworkProcess/storage/SessionStorageManager.h:
* Source/WebKit/NetworkProcess/storage/StorageAreaBase.cpp:
(WebKit::StorageAreaBase::hasListeners const): Deleted.
* Source/WebKit/NetworkProcess/storage/StorageAreaBase.h:
(WebKit::StorageAreaBase::hasListeners const):

Canonical link: https://commits.webkit.org/285106@main
